crates/store/src/state.rs

```diff
@@ -229,63 +242,48 @@ impl State {
             (account_tree, chain_mmr, nullifier_tree, notes)
         };
 
-        let block_num = block.header().block_num();
-        let block_hash = block.hash();
-        let block_data = block.to_bytes();
-
-        // signals the transaction is ready to be committed, and the write lock can be acquired
+        // Signals the transaction is ready to be committed, and the write lock can be acquired
         let (allow_acquire, acquired_allowed) = oneshot::channel::<()>();
-        // signals the write lock has been acquired, and the transaction can be committed
+        // Signals the write lock has been acquired, and the transaction can be committed
         let (inform_acquire_done, acquire_done) = oneshot::channel::<()>();
 
         // The DB and in-memory state updates need to be synchronized and are partially
         // overlapping. Namely, the DB transaction only proceeds after this task acquires the
         // in-memory write lock. This requires the DB update to run concurrently, so a new task is
         // spawned.
         let db = self.db.clone();
-        let handle =
+        let db_update_task =
             tokio::spawn(
                 async move { db.apply_block(allow_acquire, acquire_done, block, notes).await },
             );
 
+        // Wait for the message from the DB update task, that we ready to commit the DB transaction
         acquired_allowed
             .await
             .map_err(ApplyBlockError::BlockApplyingBrokenBecauseOfClosedChannel)?;
 
-        // scope to update the in-memory data
+        // Awaiting the block saving task to complete without errors
+        block_save_task.await??;
+
+        // Scope to update the in-memory data
         {
+            // We need to hold the write lock here, so that the in-memory state is not changed
             let mut inner = self.inner.write().await;
```

Comment:
I'm not sure the comment here is entirely correct: we hold the lock not so much so that in-memory state is not changed (this would be guaranteed by locking the `writer` field on line 112) but because we want to prevent inconsistent reads. That is, if we commit the DB transaction before acquiring a lock, a read request to the store could read the old in-memory state but new database state.

Reply:
Yes, you're right, thank you for noticing!